scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
-  {TimePeriodMs {:suggested-matches {HostInfo [s/Uuid]}}})
+   {0 {:suggested-matches [#uuid \"1db521d2-b1d5-41b4-9a93-a7e12abd940b\"]}
+    60000 {:suggested-matches [#uuid \"724521d2-b1d5-41b4-9a93-a7e12abd940b\"]}}"
+  {TimePeriodMs {:suggested-matches [s/Uuid]}})
 
 
 (defprotocol Optimizer
   "Protocol defining a tool to produce a schedule to execute"
-  (produce-schedule [this queue running available host-infos]
-                    "Returns a schedule of what jobs to place on what machines
-                     and what hosts to purchase at different time steps.
-                     Conforms to the Schedule schema above
-
-                     Parameters:
-                     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
-                     running -- Set of tasks running (see cook.mesos.schema instance)
-                     available -- Set of offers outstanding
-                     host-infos -- Host infos from HostFeed
-
-                     Returns:
-                     A schedule: a map where the keys are milliseconds in the future
-                     and  the values are recommendations to take at that point in the future"))
-
-(defn create-dummy-host-feed
-  "Returns an instance of HostFeed which returns an empty list of hosts"
-  [_]
-  (log/info "Creating dummy host feed")
-  (reify HostFeed
-    (get-available-host-info [this]
-      [])))
+  (produce-schedule
+    [this queue running]
+    "Returns a schedule of what jobs to place on what machines
+     and what hosts to purchase at different time steps.
+     Conforms to the Schedule schema above
+
+     Parameters:
+     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
+     running -- Set of tasks running (see cook.mesos.schema instance)
+     host-infos -- Host infos from HostFeed
+
+     Returns:
+     A schedule: a map where the keys are milliseconds in the future
+     and  the values are recommendations to take at that point in the future"))
 
 (defn create-dummy-optimizer
   "Returns an instance of Optimizer which returns an empty schedule"
   [_]
   (log/info "Creating dummy optimizer")
   (reify Optimizer
-    (produce-schedule [this queue running available host-infos]
-      {0 {:suggested-matches {}}})))
+    (produce-schedule [_ _ _] {0 {:suggested-matches []}})))
 
-(defn optimizer-cycle!
+(defn instance-runtime
+  "Returns the instance runtime."
+  [instance now]
+  (let [start (:instance/start-time instance)
+        end (:instance/end-time instance now)]
+    (- (.getTime end)
+       (.getTime start))))
+
+(defn calculate-expected-complete-time
+  "Returns the expected time remaining to job completion.
+   Simple model for if run time exceeds expected runtime: the more it exceeds expected, the longer we expect it to take.
+   TODO: get better model!"
+  [instance expected-duration now]
+  (->> (instance-runtime instance now)
+       (- expected-duration)
+       Math/abs))
```

Comment:
Why would we want `Math/abs` here? It seems like `(max 0)` would make more sense; i.e., if the time remaining is negative, then it's already done, so no time is left.

Reply:
I think this warrants a comment.